### ISSUE TITLE
Big integer output wrong with pure PHP encoder

### ISSUE THREAD

**Issue by ElfSundae** [tuupola/base62#1]
The `characters` option can be others like `1234567890`?

**Comment by tuupola** [comment]
It needs to be 62 unique characters.

**Comment by ElfSundae** [comment]
I see, thanks.

And what is the largest number when encoding and decoding with `$integer = true`?
I tried some values, it looks like the max number is about `PHP_INT_MAX / 5.5`

```php
>>> PHP_VERSION
=> "7.1.8"
>>> function_exists("gmp_init")
=> true
>>> use Tuupola\Base62Proxy
>>> $fun = function ($number) {
...     $number = (int) $number;
...     $encoded = Base62Proxy::encode($number);
...     $decoded = Base62Proxy::decode($encoded, true);
...     $same = $number === $decoded;
...     return compact('number', 'encoded', 'decoded', 'same');
... }
=> Closure {#776
     parameters: {
       $number: {},
     },
   }
>>> 
>>> $fun(123456)
=> [
     "number" => 123456,
     "encoded" => "W7E",
     "decoded" => 123456,
     "same" => true,
   ]
>>> $fun(PHP_INT_MAX)
=> [
     "number" => 9223372036854775807,
     "encoded" => "AzL8n0Y58i7",
     "decoded" => 9223372036854775559,
     "same" => false,
   ]
>>> $fun(PHP_INT_MAX/6)
=> [
     "number" => 1537228672809129216,
     "encoded" => "1pYWSpl0r6e",
     "decoded" => 1537228672809129216,
     "same" => true,
   ]
>>> $fun(PHP_INT_MAX/5)
=> [
     "number" => 1844674407370955264,
     "encoded" => "2CGcxOuPpMg",
     "decoded" => 1844674407370955202,
     "same" => false,
   ]
>>> $fun(PHP_INT_MAX/5.5)
=> [
     "number" => 1676976733973595648,
     "encoded" => "1zsZPopHqKm",
     "decoded" => 1676976733973595648,
     "same" => true,
   ]
>>> $fun(PHP_INT_MAX/5.4)
=> [
     "number" => 1708031858676810240,
     "encoded" => "22AnqhjLcKA",
     "decoded" => 1708031858676810178,
     "same" => false,
   ]
>>> $fun(1686976733973595648)   
=> [
     "number" => 1686976733973595648,
     "encoded" => "20cN1QVk7Aw",
     "decoded" => 1686976733973595586,
     "same" => false,
   ]
```

Maybe this is a bug?

**Comment by ElfSundae** [comment]
I also tried `dev-master`, it is the same as `0.8.0`

**Comment by tuupola** [comment]
It seems pure PHP encoder for some reason has limit of `PHP_INT_MAX - 248`. GMP and BcMath encoders handle full PHP_INT_MAX too.

```php
print $data = PHP_INT_MAX - 248;
print "\n";

print $encoded = (new Tuupola\Base62\PhpEncoder)->encode($data);
print "\n";
print $encoded2 = (new Tuupola\Base62\GmpEncoder)->encode($data);
print "\n";
print $encoded3 = (new Tuupola\Base62\BcmathEncoder)->encode($data);
print "\n";
print $encoded4 = (new Tuupola\Base62)->encode($data);
print "\n";

print (new Tuupola\Base62\PhpEncoder)->decode($encoded, true);
print "\n";
print (new Tuupola\Base62\GmpEncoder)->decode($encoded2, true);
print "\n";
print (new Tuupola\Base62\BcmathEncoder)->decode($encoded3, true);
print "\n";
print (new Tuupola\Base62)->decode($encoded3, true);
print "\n";

/*
9223372036854775559
AzL8n0Y58i7
AzL8n0Y58i7
AzL8n0Y58i7
AzL8n0Y58i7
9223372036854775559
9223372036854775559
9223372036854775559
9223372036854775559
*/
```

**Comment by tuupola** [comment]
It seems pure PHP encoder seems to fails with some numbers, such as 1844674407370955264.

**Comment by ElfSundae** [comment]
> Autoselecting always defaulted to PHP encoder even when GMP was available. This is fixed now in master.

But when to autoselect `BcmathEncoder` ? Maybe the fixer should be like:

```php
if (extension_loaded('gmp')) {
    $this->encoder = new Base62\GmpEncoder($this->options);
} elseif (extension_loaded('bcmath')) {
    $this->encoder = new Base62\BcmathEncoder($this->options);
} else {
    $this->encoder = new Base62\PhpEncoder($this->options);
}
```

**Comment by ElfSundae** [comment]
I tested more. Yes, GMP and BCMath are OK.

Sometimes, there is an orderly error using PhpEncoder, like:

:heavy_check_mark: `PHP_INT_MAX - 255`
:heavy_multiplication_x: `PHP_INT_MAX - 255 - (1 ... 61)`
:heavy_check_mark: `PHP_INT_MAX - 255 - 62`
:heavy_multiplication_x: `PHP_INT_MAX - 255 - 62 - (1 ... 61)`
:heavy_check_mark: `PHP_INT_MAX - 255 - 62 - 62`
:heavy_multiplication_x: `PHP_INT_MAX - 255 - 62 - 62 - (1 ... 61)`

**Comment by ElfSundae** [comment]
The key problem is this line:
```php
$digit = (integer) ($accumulator / $target_base);
```
division between big integer can not be trusted. In my system, it is only safe if the integer number has a length less than 16.

So this issue is because of the PHP float precision, I suggest PhpEncoder should be removed.

**Comment by tuupola** [comment]
Pure PHP encoder will be fixed instead. BCMath encoder is much slower than the PHP encoder so there is no point autodetecting it. User can still instantiate it manually if there is a reason to use it. BCMath encoder is included mostly for curiosity.

**Comment by tuupola** [comment]
Fixed in `0.9.0`. Thanks for the heads up!

**Comment by ElfSundae** [comment]
Well done 🎉
